perception/sandbox/polygrasp/src/polygrasp/robot_interface.py

```diff
@@ -0,0 +1,245 @@
+import torch
+
+import hydra
+import graspnetAPI
+import polymetis
+
+import ikpy.chain
+import tempfile
+
+from . import compute_quat_dist
+
+log = logging.getLogger(__name__)
+
+
+def compute_des_pose(best_grasp):
+    """Convert between GraspNet coordinates to robot coordinates."""
+
+    # Grasp point
+    grasp_point = best_grasp.translation
+
+    # Compute plane of rotation through three orthogonal vectors on plane of rotation
+    grasp_approach_delta = best_grasp.rotation_matrix @ np.array([-0.3, 0.0, 0])
+    grasp_approach_delta_plus = best_grasp.rotation_matrix @ np.array([-0.3, 0.1, 0])
+    grasp_approach_delta_minus = best_grasp.rotation_matrix @ np.array([-0.3, -0.1, 0])
+    bx = -grasp_approach_delta
+    by = grasp_approach_delta_plus - grasp_approach_delta_minus
+    bx = bx / np.linalg.norm(bx)
+    by = by / np.linalg.norm(by)
+    bz = np.cross(bx, by)
+    plane_rot = R.from_matrix(np.vstack([bx, by, bz]).T)
+
+    # Convert between GraspNet neutral orientation to robot neutral orientation
+    des_ori = plane_rot * R.from_euler("y", 90, degrees=True)
+    des_ori_quat = des_ori.as_quat()
+
+    return grasp_point, grasp_approach_delta, des_ori_quat
+
+
+def grasp_to_pose(grasp: graspnetAPI.Grasp):
+    return grasp.translation, R.from_matrix(grasp.rotation_matrix).as_quat()
+
+
+def min_dist_grasp(default_ee_quat, grasps):
+    """Find the grasp with minimum orientation distance to the reference grasp"""
+    with torch.no_grad():
+        rots_as_quat = [
+            torch.Tensor(R.from_matrix(grasp.rotation_matrix).as_quat())
+            for grasp in grasps
+        ]
+        dists = [compute_quat_dist(rot, default_ee_quat) for rot in rots_as_quat]
+        i = torch.argmin(torch.Tensor(dists)).item()
+    log.info(f"Grasp {i} has angle {dists[i]} from reference.")
+    return grasps[i], i
+
+
+def min_dist_grasp_no_z(default_ee_quat, grasps):
+    """
+    Find the grasp with minimum orientation distance to the reference grasp
+    disregarding orientation about z axis
+    """
+    with torch.no_grad():
+        rots_as_R = [R.from_quat(compute_des_pose(grasp)[2]) for grasp in grasps]
+        default_r = R.from_quat(default_ee_quat)
+        dists = [
+            np.linalg.norm((rot.inv() * default_r).as_rotvec()[:2]) for rot in rots_as_R
+        ]
+        i = torch.argmin(torch.Tensor(dists)).item()
+    log.info(f"Grasp {i} has angle {dists[i]} from reference.")
+    return grasps[i], i
+
+
+class GraspingRobotInterface(polymetis.RobotInterface):
+    def __init__(
+        self,
+        gripper: polymetis.GripperInterface,
+        k_approach=1.5,
+        k_grasp=0.72,
+        gripper_max_width=0.085,
+        # ikpy params:
+        base_elements=["panda_link0"],
+        soft_limits=[
+            (-2.70, 2.70),
+            (-1.56, 1.56),
+            (-2.7, 2.7),
+            (-2.87, -0.07),
+            (-2.7, 2.7),
+            (-0.02, 3.55),
+            (-2.7, 2.7),
+        ],
```

Comment:
Even though `GraspingRobotInterface` probably only gets initialized once in every use case, it's probably still not a good idea to have mutables as default arguments. Should default to `None` and assign values afterwards.

Reply:
Or just use tuple?